### PR TITLE
Add a way to test MAO changes in CI

### PR DESCRIPTION
**Pull request by sadasu**
These changes make it possible to test MAO changes that are
in-flight on the dev-scripts CI.

### CODE REVIEW

**Comment by sadasu** [comment]
machine-api-operator-roles.yaml can be removed  (probably as part of https://github.com/openshift-metal3/dev-scripts/pull/799) after https://github.com/openshift/machine-api-operator/pull/389 merges.

File: run-custom-mao.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash -xe
+
+# This script needs an machine-api-operator image as an input parameter
+# For example: ./run-custom-mao.sh quay.io/sdasu/machine-api-operator:sdasu-fix
+export SCRIPTS_DIR=$(pwd)
+export CUSTOM_MAO_BUILD=$1
+
+kubectl get secrets --namespace openshift-machine-api | grep mariadb
```
Comment (russellb) [inline run-custom-mao.sh]: left over debug?-

File: run-custom-mao.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+# Stop CVO, MAO, machine-api-controller and Metal3 deployments
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig scale deployment --replicas=0 cluster-version-operator --namespace openshift-cluster-version
```
Comment (russellb) [inline run-custom-mao.sh]: this config path will not always be correct

File: run-custom-mao.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig scale deployment -n openshift-machine-api --replicas=0 machine-api-operator
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig scale deployment --replicas=0 machine-api-controllers --namespace openshift-machine-api
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig scale deployment --replicas=0 metal3 --namespace openshift-machine-api
+
+REPLACEMENT_TEXT="image: ${CUSTOM_MAO_BUILD}"
+echo $REPLACEMENT_TEXT
```
Comment (russellb) [inline run-custom-mao.sh]: left over debug?

File: run-custom-mao.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+export SCRIPTS_DIR=$(pwd)
+export CUSTOM_MAO_BUILD=$1
+
+kubectl get secrets --namespace openshift-machine-api | grep mariadb
+
+# Stop CVO, MAO, machine-api-controller and Metal3 deployments
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig scale deployment --replicas=0 cluster-version-operator --namespace openshift-cluster-version
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig scale deployment -n openshift-machine-api --replicas=0 machine-api-operator
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig scale deployment --replicas=0 machine-api-controllers --namespace openshift-machine-api
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig scale deployment --replicas=0 metal3 --namespace openshift-machine-api
```
Comment (russellb) [inline run-custom-mao.sh]: why does metal3 need to be scaled down?  When the MAO restarts, either the deployment is the same and nothing needs to happen, or if it changes, it will get automatically restarted.

File: run-custom-mao.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+REPLACEMENT_TEXT="image: ${CUSTOM_MAO_BUILD}"
+echo $REPLACEMENT_TEXT
+
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig get deployments -n openshift-machine-api machine-api-operator -o yaml > mao.yaml
+sed -i -e "s!image: .*!${REPLACEMENT_TEXT}!g" mao.yaml
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig apply -f mao.yaml
```
Comment (russellb) [inline run-custom-mao.sh]: I think you can do these 3 lines in one command using `oc patch`

File: run-custom-mao.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+# Give the machine-api-operator rbac access to create and read secrets
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig apply -f machine-api-operator-roles.yaml
```
Comment (russellb) [inline run-custom-mao.sh]: Instead of having this in dev-scripts, how about making `run-custom-mao` apply the latest manifests from the MAO based on the branch it's using?  It seems like this could make CI work, but it would break in a normal deployment if it's missing some required rbac?

File: run-custom-mao.sh [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+oc --config /home/stack/dev-scripts/ocp/auth/kubeconfig scale deployment -n openshift-machine-api --replicas=1 machine-api-operator
```
Comment (russellb) [inline run-custom-mao.sh]: I don't think you need to scale down/up the MAO.  Once you stop CVO and patch the MAO deployment, it's going to restart automatically.

File: run_ci.sh [hunk 8]
```diff
@@ -169,3 +169,7 @@ for FILE in $FILESTOCACHE ; do
         sudo cp -p $FILE $cached
     fi
 done
+
+# Now that the deployment is up, replace the machine-api-operator image with
+# one that needs to be tested in this CI.
+./run-custom-mao.sh quay.io/sdasu/machine-api-operator:sdasu-fix
```
Comment (russellb) [inline run_ci.sh]: Can you put this part in a separate commit?  We'd drop this part before merging the rest.

Comment (sadasu) [reply run_ci.sh]: Yes, will do right after CI completes showing that the MAO commit linked here, works.

File: run-custom-mao.sh (hunk 1 above)
Comment (sadasu) [reply run-custom-mao.sh]: Added this and "echo $REPLACEMENT_TEXT" so that the CI console logs will have this information that will help if there are any potential issues. For example, right now this command will show that "mariadb-password" exists. When this script is run as part of changes to remove the mariadb-password creation in dev-scripts, this should show that no secrets with "mariadb" exist. I think it has value in the short term, but can be removed later.

File: run-custom-mao.sh (hunk 2 above)
Comment (sadasu) [reply run-custom-mao.sh]: True. Let me fix that.

File: run-custom-mao.sh (hunk 3 above)
Comment (sadasu) [reply run-custom-mao.sh]: See comment above. I am not opposed to removing it right away if this is viewed as noise.

File: run-custom-mao.sh (hunk 5 above)
Comment (sadasu) [reply run-custom-mao.sh]: Let me give that a try.

File: run-custom-mao.sh (hunk 6 above)
Comment (sadasu) [reply run-custom-mao.sh]: Agreed. We could make this script apply all manifests from a given branch to make it more generic.

File: run-custom-mao.sh (hunk 7 above)
Comment (sadasu) [reply run-custom-mao.sh]: Yes. I was thinking about that. By forcing a scale up right away we can save some time (time for the reconcile method to run again and discover that the MAO image has changed and then bring it down and up again with the new image) during our CI runs. I am open to removing this and let the reconcile do its thing.

File: run-custom-mao.sh (hunk 4 above)
Comment (sadasu) [reply run-custom-mao.sh]: We are not really changing anything inside Metal3's deployment object. We are just adding a Secret type to the openshift-machine-api's namespace. I am not sure if that will result in the Metal3 being restart automatically. I am trying this on my setup.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1148/

File: run-custom-mao.sh (hunk 5 above)
Comment (sadasu) [reply run-custom-mao.sh]: It appears that we can "oc patch" only the pod and not the deployment. So, to figure out the machine-api-operator pod name would be another oc (oc get pods) call. And then extract the name of the pod and plug it into the "oc patch" commands results in a 3 step operation too (can be expressed in 2 lines of code). Still want me to change to "oc patch"?

File: run-custom-mao.sh (hunk 4 above)
Comment (sadasu) [reply run-custom-mao.sh]: What I noticed was that the when MAO restarted, it started deploying another Metal3 and the old Metal3 deployment was automatically terminated. So, removing lines that scale the Metal3 and machine-api-controller replicas to 0.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1149/

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/1150/

File: run_ci.sh (hunk 8 above)
Comment (sadasu) [reply run_ci.sh]: @russellb CI has completed successfully. Is this a good time to remove the changes made to run_ci.sh?

File: run-custom-mao.sh (hunk 7 above)
Comment (sadasu) [reply run-custom-mao.sh]: I tried this yesterday. "oc patch" can be used to update a pod's image and that would automatically restart the pod with the image. The only issue I had with using "oc patch" was that I could not get it to work using the image name to be replaced as a variable. Also, that still needed the same number of steps as updating the deployment object as I have now. (1. oc get pods 2. Grab the machine-api-operator pod name 3. Update the image name within the pod)
So, continuing to use the existing approach to update the deployment. When we update the deployment object, then we specifically need to scale machine-api-operater down and up for it to pick up the new image.

File: run-custom-mao.sh [hunk 9]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash -xe
+
+# This script needs an machine-api-operator image as an input parameter
+# For example: ./run-custom-mao.sh quay.io/sdasu/machine-api-operator:sdasu-fix
+export SCRIPTS_DIR=$(pwd)
```
Comment (hardys) [inline run-custom-mao.sh]: This won't be a useful value, since the value will change depending on where the calling shell is running - you can use SCRIPTDIR as defined in common.sh (or copy the same approach).

This isn't currently used, but probably should be instead of the relative paths below?

File: run-custom-mao.sh [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+export CUSTOM_MAO_BUILD=$1
```
Comment (hardys) [inline run-custom-mao.sh]: We should validate here and exit with usage info if no CLI args are passed?

File: run-custom-mao.sh [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+export KUBECONFIG=ocp/auth/kubeconfig
+REPLACEMENT_TEXT="image: ${CUSTOM_MAO_BUILD}"
+MAO_REPO="raw.githubusercontent.com/sadasu/machine-api-operator"
```
Comment (hardys) [inline run-custom-mao.sh]: This should be configurable either via the CLI or by making the variable overridable?

File: run-custom-mao.sh [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+MAO_BRANCH="sdasu-fix"
```
Comment (hardys) [inline run-custom-mao.sh]: Same we should make this configurable, or required in the CLI args

File: run-custom-mao.sh [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+MAO_RBAC_MANIFEST="install/0000_30_machine-api-operator_09_rbac.yaml"
+
+oc --config ${KUBECONFIG} get secrets --namespace openshift-machine-api | grep mariadb
+
+# Stop CVO, MAO, machine-api-controller and Metal3 deployments
+oc --config ${KUBECONFIG} scale deployment --replicas=0 cluster-version-operator --namespace openshift-cluster-version
+oc --config ${KUBECONFIG} scale deployment --replicas=0 machine-api-operator --namespace openshift-machine-api
+
+wget https://${MAO_REPO}/${MAO_BRANCH}/${MAO_RBAC_MANIFEST} -O machine-api-operator-roles.yaml
```
Comment (hardys) [inline run-custom-mao.sh]: we should probably put this downloaded file into the ocp dir so it's cleaned on `make clean` ?

File: run-custom-mao.sh [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+# Give the machine-api-operator rbac access to create and read secrets
+oc --config ${KUBECONFIG} apply -f machine-api-operator-roles.yaml
+
+oc --config ${KUBECONFIG} get deployments -n openshift-machine-api machine-api-operator -o yaml > mao.yaml
```
Comment (hardys) [inline run-custom-mao.sh]: we should put the downloaded mao.yaml into the ocp dir?

**Comment by hardys** [comment]
A few minor nits but looks good - can you check if some of the mentioned variables can either be removed or made configurable, then this can probably go in and we'll iterate on it with any further cleanups?

File: run_ci.sh (hunk 8 above)
Comment (russellb) [reply run_ci.sh]: @sadasu Is anything doing validation to ensure that the new custom MAO comes up correctly and is operational?  It doesn't look like it ...

A suggestion:  Move where this runs to script 11 (register hosts), and run it before the worker machine set is scaled up from 0.  You can have it only run if some config variable is set pointing to a custom repo/branch of the MAO to run.  If you do this, when CI does its validation to ensure a worker comes up, you know that the worker only came up using the restarted custom MAO.

**Comment by sadasu** [comment]
Marking this a WIP while I try out some changes.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1153/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1154/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1155/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1156/

File: run-custom-mao.sh (hunk 14 above)
Comment (sadasu) [reply run-custom-mao.sh]: Done.

File: run-custom-mao.sh (hunk 13 above)
Comment (sadasu) [reply run-custom-mao.sh]: Done.

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/1158/

File: run_ci.sh [hunk 15]
```diff
@@ -140,6 +140,9 @@ for PROJ in installer ; do
     mv /home/notstack/$PROJ $HOME/go/src/github.com/$GITHUB_ORGANIZATION
 done
 
+# If you wish the deployment to install a custom MAO then set the following env variable
+export TEST_CUSTOM_MAO=true
```
Comment (russellb) [inline run_ci.sh]: Can you document this in config_example.sh / README.md ?

how about variables for the repo and branch, as well?

**Comment by russellb** [comment]
CI passed, great!

It looks like we could merge this once you drop the bits that were turning this on for CI purposes.

File: run_ci.sh (hunk 15 above)
Comment (sadasu) [reply run_ci.sh]: Yeah, the image repo and branch can be passed in via CLI set and env variables or use defaults. Will update documentation.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1159/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1160/

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/1161/

File: 11_register_hosts.sh [hunk 16]
```diff
@@ -40,6 +40,12 @@ function make_bm_workers() {
     done
 }
 
+if [ "$TEST_CUSTOM_MAO" = true ]; then
+  # Now that the deployment is up, replace the machine-api-operator image with
+  # one that needs to be tested in this CI before we bring up the workers.
+  ./run-custom-mao.sh quay.io/sdasu/machine-api-operator:sdasu-fix sadasu sdasu-fix
```
Comment (hardys) [inline 11_register_hosts.sh]: This should use `${SCRIPTDIR}` instead of a relative path, and also you need to remove the now-optional arguments here I think?

File: README.md [hunk 17]
```diff
@@ -200,3 +200,18 @@ If you see `ftype=1` then you have d_type support.
...
+with a custom image.
+- Turn on this functionality by setting TEST_CUSTOM_MAO to true.
+
+- Call ./run_custom_mao.sh from within 11_register_hosts.sh providing it with
+its parameters (use place holder already provided in 11_register_hosts.sh).
+
+For example:
+./run-custom-mao.sh <path in quay.io for the custom MAO image with tag> <repo name> <branch name>
+
+- All input parameters can be set via CUSTOM_MAO_IMAGE, REPO_NAME and MAO_BRANCH respectively.
```
Comment (hardys) [inline README.md]: To make this clearer should we say "Alternatively, all input parameters can be set via variables ..." ?

File: run-custom-mao.sh [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+    echo "Usage: ./run-custom-mao.sh <machine-api-operator image> <repo name> <branch name>"
+    echo "Machine-api-operator image to be tested is mandatory."
+    echo "repo_name and branch_name are optional parameters."
+    echo "If these are not provided, they will default to "openshift" and "master" respectively."
+    echo "Example: ./run-custom-mao.sh quay.io/sdasu/machine-api-operator:sdasu-fix sadasu sdasu-fix"
+    echo "Input parameters can also be provided via CUSTOM_MAO_IMAGE, REPO_NAME and MAO_BRANCH resp."
+    echo "Also, this script assumes that it is run after a successful dev-scripts install."
+    exit 1
+  else
+  CUSTOM_MAO_IMAGE=$1
```
Comment (hardys) [inline run-custom-mao.sh]: nit missing indentation here

File: run_ci.sh [hunk 19]
```diff
@@ -140,6 +140,9 @@ for PROJ in installer ; do
     mv /home/notstack/$PROJ $HOME/go/src/github.com/$GITHUB_ORGANIZATION
 done
 
+# If you wish the deployment to install a custom MAO then set the following to true
+export TEST_CUSTOM_MAO=false
```
Comment (hardys) [inline run_ci.sh]: I think we need to default this somewhere else, or the variable will be undefined for existing users running the 11 script?

Maybe move this to common.sh?

**Comment by sadasu** [comment]
Took care of review comments from @russellb and @hardys. But want CI to run the custom MAO test on more time. So, leaving TEST_CUSTOM_MAO to true in run_ci.sh. After another successful run, I will remove that line from run_ci.sh.

File: common.sh [hunk 20]
```diff
@@ -164,3 +164,6 @@ if [ ! -d "$IRONIC_DATA_DIR" ]; then
   sudo chown "${USER}:${USER}" "$IRONIC_DATA_DIR"
   chmod 755 "$IRONIC_DATA_DIR"
 fi
+
+# Defaults the variable to enable testing a custom machine-api-operator image
+export TEST_CUSTOM_MAO=false
```
Comment (hardys) [inline common.sh]: You need to set the false only as a default, e.g:

```
export TEST_CUSTOM_MAO=${TEST_CUSTOM_MAO:-false}
```

**Comment by hardys** [comment]
> TEST_CUSTOM_MAO to true in run_ci.sh.

Note this will be ignored in the latest revision due to the issue mentioned in `common.sh`

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/1162/

**Comment by sadasu** [comment]
After the successful CI run, now disabling testing of custom MAO image.

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1164/

File: 11_register_hosts.sh [hunk 21]
```diff
@@ -40,6 +40,16 @@ function make_bm_workers() {
     done
 }
 
+if [ "$TEST_CUSTOM_MAO" = true ]; then
+  # Now that the deployment is up, replace the machine-api-operator image with
+  # one that needs to be tested in this CI before we bring up the workers.
+  #./run-custom-mao.sh <custom mao image name> <mao repo name> <mao branch name>
+  export CUSTOM_MAO_IMAGE="quay.io/sdasu/machine-api-operator:sdasu-fix"
+  export REPO_NAME="sadasu"
+  export MAO_BRANCH="sdasu-fix"
```
Comment (hardys) [inline 11_register_hosts.sh]: These need to be removed now so that others can override via their config

File: run-custom-mao.sh [hunk 22]
```diff
@@ -0,0 +1,106 @@
+#!/bin/bash -xe
+
+if [ -z "$CUSTOM_MAO_IMAGE" ]; then
+  # Custom MAO image not already provided
+
+  if [ -z "${1}" ]; then
```
Comment (hardys) [inline run-custom-mao.sh]: A more compact way of doing this would be to default to $1, e.g

```
CUSTOM_MAO_IMAGE=${CUSTOM_MAO_IMAGE:-$1}
if [ -z "$CUSTOM_MAO_IMAGE" ]; then
...
```

File: run_ci.sh [hunk 23]
```diff
@@ -140,6 +140,10 @@ for PROJ in installer ; do
     mv /home/notstack/$PROJ $HOME/go/src/github.com/$GITHUB_ORGANIZATION
 done
 
+# If you wish the deployment to install a custom MAO then set the following to true
+# Make sure to specify the custom MAO image in 11_register_hosts.sh
```
Comment (hardys) [inline run_ci.sh]: The config shouldn't be specified in 11_register_hosts.sh, it should be provided in the environment, e.g for CI here, and for local testing via `config_$USER.sh`

File: config_example.sh [hunk 24]
```diff
@@ -53,3 +53,15 @@ set -x
...
+# Enable testing of custom machine-api-operator-image
+#export TEST_CUSTOM_MAO=true
+
+# Custom machine-api-operator image with tag
+#export CUSTOM_MAO_IMAGE="quay.io/mao-user/machine-api-operator:mao-fix"
+
+# Repository that is holding any custom machine-api-operator changes
+#export REPO_NAME="mao-user"
+
+# Name os branch in the above repo which contains the custom MAO changes
```
Comment (hardys) [inline config_example.sh]: Typo s/os/of

Also should we say "Git repository" here to differentiate from the image repository above?

File: run-custom-mao.sh [hunk 25]
```diff
@@ -0,0 +1,74 @@
+#!/bin/bash -xe
+
+CUSTOM_MAO_IMAGE=${CUSTOM_MAO_IMAGE:-$1}
+if [ -z "$CUSTOM_MAO_IMAGE" ]; then
+    # Custom MAO image not already provided
+    echo "Usage: ./run-custom-mao.sh <machine-api-operator image> <repo name> <branch name>"
+    echo "Machine-api-operator image to be tested is mandatory."
+    echo "repo_name and branch_name are optional parameters."
+    echo "If these are not provided, they will default to "openshift" and "master" respectively."
```
Comment (hardys) [inline run-custom-mao.sh]: If you want the quotes around openshift/master to be displayed in the output they need to be escaped.

**Comment by hardys** [comment]
@sadasu I made a couple of minor changes to the usage/readme and removed your test changes from CI, so I think this is ready to merge now?

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1166/

**Comment by metal3ci** [comment]
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/1170/

**Comment by metal3ci** [comment]
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/1173/
